server/routes/user_router.go

```diff
@@ -0,0 +1,229 @@
+package routes
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/rawdaGastan/grid3_auto_deployer/models"
+)
+
+type SignUpInput struct {
+	Name            string `json:"name" binding:"required"`
+	Email           string `json:"email" gorm:"unique" binding:"required"`
+	Password        string `json:"password" binding:"required"`
+	ConfirmPassword string `json:"confirmPassword" binding:"required"`
+}
+
+type AuthData struct {
+	Email string `json:"email"`
+	Code  int    `json:"code"`
+}
+
+type ChangePassword struct {
+	Email           string `json:"email"`
+	Password        string `json:"password" binding:"required"`
+	ConfirmPassword string `json:"confirmPassword" binding:"required"`
+}
+
+func (router *Router) SignUpHandler(w http.ResponseWriter, r *http.Request) {
+	signUp := SignUpInput{}
+	err := json.NewDecoder(r.Body).Decode(&signUp)
+	if err != nil {
+		errJSON, _ := json.Marshal(ErrorMsg{Message: err.Error()})
+		http.Error(w, string(errJSON), http.StatusInternalServerError)
+		return
+	}
+
+	if signUp.Password != signUp.ConfirmPassword {
+		http.Error(w, "Password and Confirm Password dosen't match", http.StatusInternalServerError)
+		return
+	}
+
+	u := models.User{
+		Name:     signUp.Name,
+		Email:    signUp.Email,
+		Password: signUp.Password,
+	}
+
+	hashedPassword, err := HashPassword(u.Password)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	u.Password = hashedPassword
+
+	// send verification code
+	code, err := SendMail(u.Email)
+	if err != nil {
+		errJSON, _ := json.Marshal(ErrorMsg{Message: err.Error()})
+		http.Error(w, string(errJSON), http.StatusInternalServerError)
+		return
+	}
+	u.Code = code
+	fmt.Printf("code: %v\n", code)
+	msg := "Verfification Code has been sent to " + u.Email
+	msgBytes, err := json.Marshal(msg)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	router.db.SetCache(u.Email, u)
+
+	w.WriteHeader(http.StatusCreated)
```

Comment:
add `w.Header().Set("Content-Type", "application/json")`
and you can aslo create a response writer function to be used for all responses. for example:
```
func writeRes(w http.ResponseWriter, content interface{}) {
	contentJson, err := json.Marshal(content)
	if err != nil {
		writeErrRes(w, err) // error response function i mentioned above
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	w.Write(contentJson)
}
```